src/training/scheduler.h

```diff
@@ -18,6 +18,7 @@ class Scheduler : public TrainingObserver {
   std::vector<Ptr<ValidatorBase>> validators_;
 
   bool first_{true};
+  bool endOfStdin_{false};
```

Comment:
This looks fishy. Please add a comment why this variable exists.

Reply:
It helps to properly stop the training when the data stream from stdin is over (otherwise there would be an infinite loop of `corpus->next`) and this works with (1) training for multiple epochs and (2) renewing the training. Simply `if(tsv_ && state_->epoch > 1)` does not work with (1), and relying on the difference between `state_->epochs` and `state_->prevEpochs` did not work with (2).

I tried to put this into a comment in the code.